bin/propolis-server/src/lib/spec/builder.rs

```diff
@@ -110,6 +113,36 @@ impl SpecBuilder {
         Ok(())
     }
 
+    /// Add a boot option to the boot option list of the spec under construction.
+    pub fn add_boot_option(
+        &mut self,
+        item: &BootOrderEntry,
+    ) -> Result<(), SpecBuilderError> {
+        let boot_order = self.spec.boot_order.get_or_insert(Vec::new());
+
+        use crate::spec::StorageDevice;
+        use crate::spec::{NvmeDisk, VirtioDisk};
+
+        let is_disk =
+            self.spec.disks.values().any(|disk| match &disk.device_spec {
+                StorageDevice::Virtio(VirtioDisk { backend_name, .. }) => {
+                    backend_name == item.name.as_str()
+                }
+                StorageDevice::Nvme(NvmeDisk { backend_name, .. }) => {
+                    backend_name == item.name.as_str()
+                }
+            });
```

Comment:
yeah so it turns out the reason i thought so strongly that this matching is correct is because.. it was _for phd_. it was, of course, incorrect for disk requests as parsed from Nexus. this just matching on the `device_name` as of https://github.com/oxidecomputer/propolis/pull/756/commits/ba459299fac1e58cbbc88c5b1e589f80d65fbd9a, which also swaps how the disk name is used to match `parse_disk_from_request` on both `device_name` and `backend_name`.

in the process the gross matching is now gone. and it actually works now. yay :)